### PR TITLE
added QueryBuilder interface for SelectQuery,UpdateQuery,DeleteQuery

### LINKED ISSUE #497: Setting Where condition of SelectQuery,UpdateQuery centrally
Hi All,

I wanted to filter `SelectQuery`, `UpdateQuery`, `DeleteQuery` typed queries in a single place. I've tried interfaces&switch type but always had boilerplate code. Later tried generics in go 1.18, but again got boilerplate code.

Is there a way to get rid of boilerplate code at `userFilter` function in the following example:

```go
package main

import (
	"context"
	"database/sql"
	"github.com/google/uuid"
	"github.com/uptrace/bun"
	"github.com/uptrace/bun/dialect/sqlitedialect"
	"github.com/uptrace/bun/driver/sqliteshim"
	"github.com/uptrace/bun/extra/bundebug"
)

type User struct {
	bun.BaseModel  `bun:"table:users,alias:u"`
	ID             uuid.UUID `bun:",pk,type=uuid" json:"id"`
	OrganizationID uuid.UUID `json:"organization_id,omitempty"`
	Email          string    `json:"email"`
	IsSysadm       bool      `json:"is_sysadm"`
}

type Query interface {
	~*bun.SelectQuery | ~*bun.UpdateQuery | ~*bun.DeleteQuery
}

// boilerplate code in this func, calling WhereGroup for each type
func userFilter[T Query](oID uuid.UUID, qI T) T {
	switch q := interface{}(qI).(type) {
	case *bun.SelectQuery:
		q.WhereGroup("AND", func(query *bun.SelectQuery) *bun.SelectQuery {
			return query.Where("u.organization_id = ?", oID)
		})
	case *bun.UpdateQuery:
		q.WhereGroup("AND", func(query *bun.UpdateQuery) *bun.UpdateQuery {
			return query.Where("u.organization_id = ?", oID)
		})
		q.ExcludeColumn("organization_id", "is_sysadm")
	case *bun.DeleteQuery:
		q.WhereGroup("AND", func(query *bun.DeleteQuery) *bun.DeleteQuery {
			return query.Where("u.organization_id = ?", oID)
		})
	}
	return qI
}

func main() {
	ctx := context.Background()
	sqlite, err := sql.Open(sqliteshim.ShimName, "file::memory:?cache=shared")
	if err != nil {
		panic(err)
	}
	sqlite.SetMaxOpenConns(1)
	db := bun.NewDB(sqlite, sqlitedialect.New())
	db.AddQueryHook(bundebug.NewQueryHook(
		bundebug.WithVerbose(true),
		bundebug.FromEnv("BUNDEBUG"),
	))
	if err = db.ResetModel(ctx, (*User)(nil)); err != nil {
		panic(err)
	}

	// Usage examples below:
	oID := uuid.New()
	// SelectQuery
	users := make([]User, 0)
	if err = userFilter(oID, db.NewSelect().Model(&users)).
		OrderExpr("id ASC").Scan(ctx); err != nil {
		panic(err)
	}
	// UpdateQuery
	usr := &User{ID: uuid.New(), IsSysadm: false}
	if _, err = userFilter(oID, db.NewUpdate().Model(usr)).
		Where("id = ?", usr.ID).Exec(ctx); err != nil {
		panic(err)
	}
	// DeleteQuery
	if _, err = userFilter(oID, db.NewDelete().Model((*User)(nil))).
		Where("id = ?", uuid.New()).Exec(ctx); err != nil {
		panic(err)
	}
}

```

### PR DESCRIPTION
**Pull request by derkan**
closes #497

### CODE REVIEW

File: query_base.go [hunk 1]
```diff
@@ -65,6 +65,17 @@ var (
...
+type QueryBuilder interface {
+	Query
+	Where(query string, args ...interface{}) QueryBuilder
+	WhereGroup(sep string, fn func(QueryBuilder) QueryBuilder) QueryBuilder
+	WhereOr(query string, args ...interface{}) QueryBuilder
+	WhereDeleted() QueryBuilder
+	WhereAllWithDeleted() QueryBuilder
+	WherePK(cols ...string) QueryBuilder
+	Unwrap() interface{}
+}
```
Comment (vmihailenco) [inline query_base.go:78]: Please gofmt and add compile-time checks:

```go
var _ QueryBuilder = (*selectQueryBuilder)(nil)
var _ QueryBuilder = (*updateQueryBuilder)(nil)
var _ QueryBuilder = (*deleteQueryBuilder)(nil)
```

File: query_delete.go [hunk 2]
```diff
@@ -286,3 +286,48 @@ func (q *DeleteQuery) afterDeleteHook(ctx context.Context) error {
...
+
+//------------------------------------------------------------------------------
+type deleteQueryBuilder struct {
+	*DeleteQuery
+}
+
+func (q *deleteQueryBuilder) WhereGroup(sep string, fn func(QueryBuilder) QueryBuilder) QueryBuilder {
+	q.DeleteQuery.WhereGroup(sep, func(qs *DeleteQuery) *DeleteQuery {
+		return fn(&deleteQueryBuilder{qs}).(*deleteQueryBuilder).DeleteQuery
+	})
```
Comment (vmihailenco) [inline query_delete.go:298]: This and other methods should be:

```
q.DeleteQuery = q.DeleteQuery.WhereGroup(sep, func(qs *DeleteQuery) *DeleteQuery {
	return fn(q).(*deleteQueryBuilder).DeleteQuery
})
```

**Comment by vmihailenco** [comment]
Could you also add a test to `internal/dbtest/query_test.go`?

File: query_delete.go [hunk 3]
```diff
@@ -286,3 +286,48 @@ func (q *DeleteQuery) afterDeleteHook(ctx context.Context) error {
...
 	return nil
 }
+
+//------------------------------------------------------------------------------
+type deleteQueryBuilder struct {
+	*DeleteQuery
+}
+
+func (q *deleteQueryBuilder) WhereGroup(sep string, fn func(QueryBuilder) QueryBuilder) QueryBuilder {
+	q.DeleteQuery.WhereGroup(sep, func(qs *DeleteQuery) *DeleteQuery {
```
Comment (vmihailenco) [inline query_delete.go:296]: What about saving the return value `q.DeleteQuery = q.DeleteQuery` and everywhere else? In theory, the user can return a new `*DeleteQuery` from the callback.

File: query_base.go [hunk 4]
```diff
@@ -65,6 +65,24 @@ var (
 	_ IDB = (*Tx)(nil)
 )
 
+// QueryBuilder is used SelectQuery,UpdateQuery, DeleteQuery for adding new Where conditions
```
Comment (vmihailenco) [inline query_base.go:68]: >for adding new Where conditions

Later, we can add more methods here, for example, `Model() QueryBuilder` and `Table(...string) QueryBuilder`.

For some methods that don't exist in some of the supported queries (SelectQuery/UpdateQuery/DeleteQuery), we could add a no-op/empty implementation that does nothing but implements the interface.

File: query_delete.go (hunk 3 above)
Comment (derkan) [reply query_delete.go:296]: that makes sense, didn't realized the returned value, changing

**Comment by vmihailenco** [comment]
Looks good - thanks :+1:

**Comment by frederikhors** [comment]
Is there any docs for this? @vmihailenco can we merge docs in this repo? So anyone can update the docs too?

**Pull request merged**